docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -0,0 +1,110 @@
+
+And explained how this approach should be sufficiently collision resistant:
+> Yeah, AFAIK, 20 bytes should be collision resistance when the preimages are unique and not malleable. A space of 2^160 would expect some collision to be likely around 2^80 elements (birthday paradox). And if you want to find a collision for some existing element in the database, it is still 2^160. 2^80 only is if all these elements are written to state.
+
+> The good example you brought up was eg. a public key bytes being a valid public key on two algorithms supported by the codec. Meaning if either was broken, you would break accounts even if they were secured with the safer variant. This is only as the issue when no differentiating type info is present in the preimage (before hashing into an address).
+
+> I would like to hear an argument if the 20 bytes space is an actual issue for security, as I would be happy to increase my address sizes in weave. I just figured cosmos and ethereum and bitcoin all use 20 bytes, it should be good enough. And the arguments above which made me feel it was secure. But I have not done a deeper analysis.
+
+In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
+approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
+the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+
+## Decision
+
+### Legacy Public Key Addresses Don't Change
+
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. We
+don't want to change existing addresses. So the addresses for these two key types will remain the same.
+
+The current multisig public keys use amino serialization to generate the address. We will retain
+those public keys and their address formatting, and call them "legacy amino" multisig public keys
+in protobuf. We will also create multisig public keys without amino addresses to be described below.
+
+### Canonical Public Key Addresses
+
+Following on the discussion in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we propose the
+following approach.
+
+All public key types will have a unique protobuf message type such as:
+
+```proto
+package cosmos.crypto.sr25519;
+
+message PubKey {
+  bytes key = 1;
+}
+```
+ 
+All protobuf messages have unique fully qualified names, in this example `cosmos.crypto.sr25519.PubKey`.
+These names are derived directly from .proto files in a standardized way and used
+in other places such as the type URL in `Any`s. Since there is an easy and obvious
+way to get this name for every protobuf type, we can use this message name as the
+key type prefix when creating addresses.
+
+We define the canonical address format for new (non-legacy) public keys as
+`Sha256(fmt.Sprintf("%s/%x, proto.MessageName(key), key.Bytes())[:20]`. This takes
```

Comment:
Current secp256k1 uses ripemd160 instead of [:20]. The latter is faster to compute, but are there other trade-offs (in particular, security-wise)?

Reply:
It is needed for bitcoin compatibility. There were other designs, but the 2017 fundraiser addresses were generated with a bitcoin secp256k1 address derivation and the go code had to be compatible.

But I cannot tell you if 20 bytes prefix is just as safe

Reply:
ripemd160 is an odd and basically pointless cryptographic choice and there is no reason to retain it.

It's a slow hash function and doesn't improve security in a meaningful way.

Reply:
Why not blake3 @robert-zaremba ?

Reply:
I'm all for `blake3` (especially that it's officially endorsed on the blake2 side) but in other discussion majority was afraid that it's too fresh. https://github.com/cosmos/cosmos-sdk/pull/7259#issuecomment-688896136

Reply:
I'd be in favor of blake3 if we're going that route. I think the issue regarding state sync was more that the cryptography is a relatively minor part of snapshot sync. Whereas with address formats it's more critical. So I think that would merit it taking a more central role and us attempting to choose the best algorithm, which by everything I can see appears to be blake3.

Reply:
Although, I would say that the downside that needs to be considered is the availability of a high-quality implementation. Clients should be able to compute addresses from public keys and sha256 pretty ubiquitous. Address hashing is also done over a very small piece of data so I expect the performance benefits would be minimal. It probably only makes sense if there is some security improvement to a different algorithm. I would expect we'll see more of a performance effect if we're looking at block or merkle tree hashing.

Reply:
`blake2b` is considered safer than `sha2`. Changing address creation algorithm is way more complex thing than chunks hashes.

Reply:
> `blake2b` is considered safer than `sha2`.

Can you articulate the safety benefits `blake2b` or `blake3` would offer for address hasing?

> Changing address creation algorithm is way more complex thing than chunks hashes.

Yes, I totally agree. What we define here will (hopefully) be in use for a long time.

Reply:
> BLAKE2b is faster than MD5, SHA-1, SHA-2, and SHA-3, on 64-bit x86-64 and ARM architectures.[3] BLAKE2 provides security superior to SHA-2 and similar to that of SHA-3: immunity to length extension, indifferentiability from a random oracle, etc.

source: [wikipedia](https://en.wikipedia.org/wiki/BLAKE_%28hash_function%29#BLAKE2)

Blake2 is conjectured to be random oracle indifferentiable, a stronger property that SHA-256 doesn't have, and which implies the three classic hash function properties.

AFAIK, blake3 adds parallelism to the design goals, so it's much faster.

Computing an address is something we will do often. In general, my approach is: if something is validated to work correctly and is better than we should just use it.

Reply:
So hypothetically it would be easier to find a pre-image for a SHA256 address than a Blake2 address?

I would note that Blake3 reduced the number of rounds from Blake2 to get a speed up. We should review the academic assessment of that decision because it appears a bit controversial. For something like this where performance is less critical, maybe we err on the side of caution and go with Blake2?

Reply:
> So hypothetically it would be easier to find a pre-image for a SHA256 address than a Blake2 address?

Not worse, and it's more random.

Reply:
> Blake3 reduced the number of rounds from Blake2 to get a speed up

It uses a different mechanism, so I can't say if the reduction of number of rounds impacts the security vs Blake2 or not. Probably asking one of the authors (Zooko?) would be the best approach.

Reply:
> maybe we err on the side of caution and go with Blake2

This seams to be the balanced solution. Although it would be still good to check out with zcash about their view. I will try to reach out to some of them.

Reply:
The new draft uses `blake2b`. Both because it has better library support and has been around longer and vetted more thoroughly (as I understand it).

I'm assuming `blake2b` is preferable to `blake2s` for our use case?

Reply:
BLAKE2b is designed and optimized for 64-bit CPUs. BLAKE2s is designed and optimized for 8, 16, 32-bit CPUs. 
BLAKE2 has indeed excellent Golang and Rust libraries support. Today there are very few BLAKE3 supported implementations. The speed boost observed with BLAKE3 is not really a selling point regarding ADR 028.